File: android/src/main/java/com/unact/yandexmapkit/lite/PolylineMapObjectController.java
package com.unact.yandexmapkit.lite;

import androidx.annotation.NonNull;

import com.yandex.mapkit.geometry.Point;
import com.yandex.mapkit.map.LineStyle;
import com.yandex.mapkit.map.MapObject;
import com.yandex.mapkit.map.MapObjectCollection;
import com.yandex.mapkit.map.MapObjectTapListener;
import com.yandex.mapkit.map.PolylineMapObject;

import java.lang.ref.WeakReference;
import java.util.Map;

public class PolylineMapObjectController extends MapObjectController implements MapObjectTapListener {
  public final PolylineMapObject polyline;
  private boolean consumeTapEvents = false;
  @SuppressWarnings({"UnusedDeclaration", "FieldCanBeLocal"})
  private final WeakReference<YandexMapController> controller;
  public final String id;

  @SuppressWarnings({"ConstantConditions", "unchecked"})
  public PolylineMapObjectController(
    MapObjectCollection parent,
    Map<String, Object> params,
    WeakReference<YandexMapController> controller
  ) {
    PolylineMapObject polyline = parent.addPolyline(
      UtilsLite.polylineFromJson((Map<String, Object>) params.get("polyline"))
    );

    this.polyline = polyline;
    this.id = (String) params.get("id");
    this.controller = controller;

    polyline.setUserData(this.id);
    polyline.addTapListener(this);
    update(params);
  }

  @SuppressWarnings({"ConstantConditions", "unchecked"})
  public void update(Map<String, Object> params) {
    polyline.setGeometry(UtilsLite.polylineFromJson((Map<String, Object>) params.get("polyline")));
    polyline.setZIndex(((Double) params.get("zIndex")).floatValue());
    polyline.setVisible((Boolean) params.get("isVisible"));
    polyline.setStrokeColor(((Number) params.get("strokeColor")).intValue());
    polyline.setStyle(new LineStyle(
      ((Double) params.get("strokeWidth")).floatValue(),
      ((Double) params.get("gradientLength")).floatValue(),
      ((Number) params.get("outlineColor")).intValue(),
      ((Double) params.get("outlineWidth")).floatValue(),
      (Boolean) params.get("isInnerOutlineEnabled"),
      ((Double) params.get("turnRadius")).floatValue(),
      ((Double) params.get("arcApproximationStep")).floatValue(),
      ((Double) params.get("dashLength")).floatValue(),
      ((Double) params.get("gapLength")).floatValue(),
      ((Double) params.get("dashOffset")).floatValue()
    ));

    consumeTapEvents = (Boolean) params.get("consumeTapEvents");
  }

  public void remove() {
    polyline.getParent().remove(polyline);
  }

  @Override
  public boolean onMapObjectTap(@NonNull MapObject mapObject, @NonNull Point point) {
    controller.get().mapObjectTap(id, point);

    return consumeTapEvents;
  }
}

File: ios/Classes/lite/PolylineMapObjectController.swift
import YandexMapsMobile

class PolylineMapObjectController: NSObject, MapObjectController, YMKMapObjectTapListener {
  public let polyline: YMKPolylineMapObject
  private var consumeTapEvents: Bool = false
  public weak var controller: YandexMapController?
  public let id: String

  public required init(
    parent: YMKMapObjectCollection,
    params: [String: Any],
    controller: YandexMapController
  ) {
    let polyline = parent.addPolyline(with: UtilsLite.polylineFromJson(params["polyline"] as! [String: Any]))

    self.polyline = polyline
    self.id = params["id"] as! String
    self.controller = controller

    super.init()

    polyline.userData = self.id
    polyline.addTapListener(with: self)
    update(params)
  }

  public func update(_ params: [String: Any]) {
    polyline.geometry = UtilsLite.polylineFromJson(params["polyline"] as! [String: Any])
    polyline.zIndex = (params["zIndex"] as! NSNumber).floatValue
    polyline.isVisible = (params["isVisible"] as! NSNumber).boolValue
    polyline.setStrokeColorWith(UtilsLite.uiColor(fromInt: (params["strokeColor"] as! NSNumber).int64Value))
    polyline.style = YMKLineStyle(
      strokeWidth: (params["strokeWidth"] as! NSNumber).floatValue,
      gradientLength: (params["gradientLength"] as! NSNumber).floatValue,
      outlineColor: UtilsLite.uiColor(fromInt: (params["outlineColor"] as! NSNumber).int64Value),
      outlineWidth: (params["outlineWidth"] as! NSNumber).floatValue,
      innerOutlineEnabled: (params["isInnerOutlineEnabled"] as! NSNumber).boolValue,
      turnRadius: (params["turnRadius"] as! NSNumber).floatValue,
      arcApproximationStep: (params["arcApproximationStep"] as! NSNumber).floatValue,
      dashLength: (params["dashLength"] as! NSNumber).floatValue,
      gapLength: (params["gapLength"] as! NSNumber).floatValue,
      dashOffset: (params["dashOffset"] as! NSNumber).floatValue
    )

    consumeTapEvents = (params["consumeTapEvents"] as! NSNumber).boolValue
  }

  public func remove() {
    polyline.parent.remove(with: polyline)
  }

  func onMapObjectTap(with mapObject: YMKMapObject, point: YMKPoint) -> Bool {
    controller!.mapObjectTap(id: id, point: point)

    return consumeTapEvents
  }
}
